r/vignettes/arrow.Rmd

```diff
@@ -1,227 +1,223 @@
-share data between R and Python (`pyarrow`) efficiently, enabling you to take
-advantage of the vibrant ecosystem of Python packages that build on top of
-Apache Arrow. See `vignette("python", package = "arrow")` for details.
+Along the same lines, the `$` operator can be used to extract named columns:
 
-## Access to Arrow messages, buffers, and streams
+```{r}
+dat$y
+```
 
-The `arrow` package also provides many lower-level bindings to the C++ library, which enable you
-to access and manipulate Arrow objects. You can use these to build connectors
-to other applications and services that use Arrow. One example is Spark: the
-[`sparklyr`](https://spark.rstudio.com/) package has support for using Arrow to
-move data to and from Spark, yielding [significant performance
-gains](https://arrow.apache.org/blog/2019/01/25/r-spark-improvements/).
+Note the output: individual columns in an Arrow Table are represented as Chunked Arrays, which are one-dimensional data structures in Arrow that are roughly analogous to vectors in R. 
+
+Tables are the primary way to represent rectangular data in-memory using Arrow, but they are not the only rectangular data structure used by the Arrow C++ library: there are also Datasets which are used for data stored on-disk rather than in-memory, and Record Batches which are fundamental building blocks but not typically used in data analysis. 
+
+To learn more about the different data object classes in `arrow`, see the article on [data objects](./data_objects.html).
+
+## Converting Tables to data frames
+
+Tables are a data structure used to represent rectangular data within memory allocated by the Arrow C++ library, but they can be coerced to native R data frames (or tibbles) using `as.data.frame()`
+
+```{r}
+as.data.frame(dat)
+```
+
+When this coercion takes place, each of the columns in the original Arrow Table must be converted to native R data objects. In the `dat` Table, for instance, `dat$x` is stored as the Arrow data type int32 inherited from C++, which becomes an R integer type when `as.data.frame()` is called. 
+
+In most instances the data conversion takes place automatically and without friction: a column stored as a timestamp in Arrow becomes a POSIXct vector in R, for example. However, there are some instances where the mapping between Arrow data types and R data types is not exact and care is required. As an example, Arrow does not have an analog of the POSIXlt class: date/time data expressed as POSIXlt objects will preserve the internal list structure of the POSIXlt object, but it will arrive in Arrow as a list. As a consequence users need to decide if what needs to be preserved is the timestamp (in which case it is better to coerce to POSIXlt to POSIXct before translating to Arrow), or if a POSIXlt-style list is preferable. Similar issues exist in the other direction: Arrow dictionary objects are a little more flexible than R factors, for instance. 
```

Comment:
> date/time data expressed as POSIXlt objects will preserve the internal list structure of the POSIXlt object, but it will arrive in Arrow as a list

Please can you give me a code example for this? Asking as I opened [a ticket for a bug when converting between POSIXlt and Arrow](https://issues.apache.org/jira/browse/ARROW-18263), and there is some weird stuff going on there, and I'm trying to work out how to get this POSIXlt -> list behaviour.

Reply:
Try this, maybe? 

```r
tm <- as.POSIXlt(c(Sys.time(), Sys.time()))
arrow::Array$create(tm)
```

It seems to arrive in Arrow in the expected list format. But I am not surprised that there are weird cases. For example, an R user might intuitively think you can import a single POSIXlt as a scalar type (after all, you *can* do that for a single POSIXct object), but this errors:

```r
arrow::Scalar$create(as.POSIXlt(Sys.time()))
```

I dunno. Maybe I should find a safer example than POSIXlt/POSIXct for this section? I feel like I'm skating on thin ice here because I'm not sure how much of this behaviour is desirable. Perhaps there's a better example involving the various integer types?

Reply:
I don't know how much is desirable either (we should think about why it's doing it like this, and what (if anything) the alternative might look like).  Can we relegate the "weird" cases to other vignettes and keep things relatively straightforward here?  Maybe just allude to the fact that it's more complex for POSIXlt and link to where that's mentioned, or pick a simpler example for this vignette if there is one.

Thank you for surfacing this though, it's great to see these rough edges exposed.  Mind opening up a JIRA asking if this is desired behaviour? I think this warrants more discussion. Feel free to tag me in it and/or link to this comment.

Reply:
Okay, I've opened a jira issue here: https://issues.apache.org/jira/browse/ARROW-18337

In the meantime, I've simplified this passage of text. It now makes the more generic point that it's possible to customise how data types are converted, and then points the reader at the data types vignette